rfd/0159-webasset-cache.md

```diff
@@ -0,0 +1,170 @@
+
+Ultimately, the user can setup their cleanup however they wish, and Teleport is responsible for filling up the bucket.
+
+
+
+#### Serving the missing webassets
+Currently, our static file serving looks [like this](https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L513C4-L518)
+```go
+if strings.HasPrefix(r.URL.Path, "/web/app") {
+    fs := http.FileServer(cfg.StaticFS)
+
+    fs = makeGzipHandler(fs)
+    fs = makeCacheHandler(fs)
+
+    http.StripPrefix("/web", fs).ServeHTTP(w, r)
+}
+```
+
+nothing too complicated besides setting some headers, compression, and then
+serving whatever file is requested. The change would be instead of serving,
+first check if the file exists in the embedded file system. If it does, send as
+usual. If it doesn't, ask auth for the missing webasset by name, auth downloads
+from the bucket, and then ships back to proxy to serve.
+```go
+if strings.HasPrefix(r.URL.Path, "/web/app") {
+	// do everything the same if the file exists
+	if (fileExists) {
+		fs := http.FileServer(cfg.StaticFS)
+
+		fs = makeGzipHandler(fs)
+		fs = makeCacheHandler(fs)
+
+		http.StripPrefix("/web", fs).ServeHTTP(w, r)
+        return
+	}
+	writer := &MemoryWriterAt{}
+	cfg.WebassetHandler.Download(cfg.Context, fileName, writer)
+
+	// detect content type
+	// compress
+	http.ServeContent(w, r, fileNameToSkip, time.Now(), bytes.NewReader(writer.buffer.Bytes()))
+}
+```
+
+
+### Security
+This feature should be **opt in only** via configuration. 
+
+Because the proxy relies on auth to upload/get the files from the bucket,
+customers can use their existing credential setup on the auth server with this
+service and _shouldn't_ have to change anything. proxy will not require AWS
+credentials.
+
+If someone has gained access to the bucket, they could potentially overwrite files with their own code that the
+browser will then try to download. We may be able to get around this by implementing some sort of signing to
+our webasset bundles but would not be part of phase1 of this project.
+
+#### Alternative approaches
+Another approach that was discussed that didn't involve an s3 bucket was storing all the webassets in an in-memory
+cache in the auth server. This would allow any proxy to reach out on start and transfer it's webassets over (if the version
+was different) and then any other proxy could reach out to auth to download the missing files.
+
+This had a bit more complexity because it would mean that _all_ auths would have to have _all_ in-use versions of the
+webassets to ensure any proxy could hit any auth for the missing files. The benefit of approach is everything would be
+self-contained and not rely on external storage. However, being much more complex, with many different listeners/watchers, it could lead to more issues down the road.
```

Comment:
I'm confused by this and I feel like S3 is over complicated for this approach

Could the proxy just grab the latest webasset files from the auth server, if it notices that the auth server it has connected/re-connected to does not match the proxy version, or the requested filename does not exist?

The auth servers are upgraded first, are always on the version that the proxy _should_ be displaying, and have the webassets built into the bundle

Reply:
That wouldn't help in the reverse direction. Take this example
Auth running 15
Proxy running 15
Proxy running 15

If we upgrade auth and one proxy to 16, if the 15 proxy has to serve 16 assets, it can reach out to auth and be all good.

But if the 16 proxy has to serve 15 assets, it has nowhere to ask and will fail. So we have to be able to serve _all_ current proxy versions that are running

Reply:
We recommend that all auth servers are upgraded before any proxy is upgraded though

> If several Auth Service instances are running in a high availability configuration (for example, in an AWS Auto Scaling group), you must shrink the group to just one Auth Service before performing an upgrade.

> Upgrade the Auth Service to the next major version first. If there are data format changes introduced in the new version, the Auth Service performs the necessary migrations. After the upgrade, start the Auth Service and CONFIRM that it's in a healthy state before continuing.

https://goteleport.com/docs/upgrading/overview/

> But if the 16 proxy has to serve 15 assets, it has nowhere to ask and will fail. So we have to be able to serve all current proxy versions that are running

The proxy shouldn't ever serve old assets, all older proxies that re-connect to an auth server which version doesn't match should know that they are out of date and to just grab the assets from the auth servers, to always serve the latest version available

Reply:
I suppose it could work if we just said anytime a proxy connects to auth, pull the webassets from auth and just serve those. so proxy version wouldn't matter and only the auth verison

Reply:
Yeah, that could work to simplify things further

Reply:
hmm, that would also mean that auth doesn't have to trust what proxy sends to it and reverses the trust back to auth

Reply:
Granted, it would alleviate some of the issues and make it only relevant for "sessions already open when the upgrade happens" but the s3 version would make it never happen (... unless the file has been cleaned up and this unhappy path happens before the next interval)

Reply:
Slab shows this when it realises there are new webassets. I feel like we could do the same

<img width="486" alt="image" src="https://github.com/gravitational/teleport/assets/7922109/abffa442-98c5-4075-8db2-6d0ce8a6e878">

Reply:
This was one of our original proposals and design wasn't a fan. But I can bring it up again and get more feedback

Reply:
I feel like it's an easier solution than using S3 as an intermediate cache where we have to trust what the proxies upload

Reply:
Design still isn't a fan, but it would be better than the product inexplicably breaking. Is there a more elegant solution we're missing? Slab is the only service I know of that shows a message like this with any frequency.

Reply:
Updated the solution slightly where auth is the only thing uploading its webassets and proxy merely asks auth for the missing webassets. This avoids any "proxy tells auth what to do" during the upload. However, it now adds an extra requirement where any proxy that wishes this to be enabled _must_ be of a version of auth that has uploaded its file.